### DISCUSSION TITLE
409 Conflict response "Cliente già esistente" in fase di creazione

### DISCUSSION THREAD

**Discussion by MattiaMarchiorato** [fattureincloud/api#66] [Q&A] [answered]
Buonasera community,

Vi scriviamo per sapere se qualcuno si è imbattuto nel nostro dubbio/problema:
Come viene verificato dalle API se un cliente esiste già in fase di creazione? 
Riceviamo il seguente errore:
[409] Client error: POST https://api-v2.fattureincloud.it/c/COMPANY_ID/entities/clients resulted in a 409 Conflict response:
{"error":{"message":"Cliente gi\u00e0 esistente."}}

Grazie mille

MM

**Comment by valmoz** [comment, marked as answer]
Buongiorno,
i campi su cui effettuiamo il controllo di duplicità per le entities (clients/suppliers) sono:
- code
- vat_number
- tax_code
- name

I diversi campi vengono aggiunti al controllo solo se dichiarati esplicitamente nella richiesta di creazione.

Grazie
Mauro

**Reply by MattiaMarchiorato** [reply]
Ciao Mauro,

https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/entities/clients
riguardo ai clienti come possiamo capire quali sono i dati obbligatori? Non vedo riferimenti nella documentazione.
Nel caso in cui si inserisca soltanto il name vuol dire che non possiamo avere 2 "Mario Rossi" registrati?

Grazie mille

A presto

MM

**Reply by valmoz** [reply]
Per quanto riguarda i dati obbligatori, puoi controllare gli schema all'interno dell'API Reference.
Ad esempio, nella [Create Client](https://developers.fattureincloud.it/api-reference#post-/c/-company_id-/entities/clients) cliccando su "schema" nella request body è possibile osservare che "name" e "value" sono obbligatori, in quanto sono evidenziati tramite asterisco rosso.

![image](https://user-images.githubusercontent.com/7254293/188481894-32b0074d-fecc-4813-95a7-838d4f68f2ee.png)

Per quanto riguarda il nome, confermo che non è possibile inserire due clienti con lo stesso nome, ma solo nel caso in cui l'unico dei parametri inseriti sia "name". Aggiungendo uno degli altri campi su cui effettuiamo il controllo (ad esempio il codice fiscale) tutto dovrebbe funzionare correttamente.

**Reply by MattiaMarchiorato** [reply]
Ok, forse anche il required è stato aggiunto recentemente perchè quando abbiamo sviluppato l'ingregrazione non era presente. Per quanto riguarda l'univocità del nome è un problema perchè lavoriamo spesso con clienti esteri a cui non richiediamo quindi il codice fiscale e può succedere che esistano due "Mario Rossi", forse avrebbe senso aggiungere prima del nome la verifica sull'email se presente? Perchè nel nostro caso l'abbiamo e spesso è un identificativo univoco, cosa ne pensi? Oltre al fatto che come immagini chiedere il codice fiscale ad un privato (rispetto chiedere l'email che probabilmente una piattafoma web ha già per l'utente) è abbastanza noioso e sbarrante.

**Reply by valmoz** [reply]
Buongiorno Mattia.
Senza andare a modificare il comportamento attuale delle API, potete tranquillamente utilizzare il campo "code".
Assegnando ai due utenti con lo stesso nome un code univoco, non ci saranno problemi di duplicazione.

Grazie
Mauro

**Reply by MattiaMarchiorato** [reply]
Ciao Mauro,

Si recentemente abbiamo aggiunto proprio il code prendendo l'id dell'utente sulla nostra piattaforma, non credo sia la soluzione ottimale ma ce lo faremo andare bene così.

A presto

MM

**Reply by valmoz** [reply]
Ok, in caso di ulteriori dubbi scrivi pure senza problemi.

Grazie
Mauro
